Email producers with order/delivery instructions #275

pmackay opened the pull request:
Would be great to get a review of this as a sanity check.

RohanM commented:
Hi @pmackay ,

I noticed that you assigned the "aus review" label. Are you looking for further review or have you made those changes somewhere I can't see?

pmackay commented:
@RohanM sorry for confusion - the idea of a tag was something we discussed with Kirsten on Monday, and I added it to this, but had not yet pushed 67b17de695ae000dc81501a0440a1b6e42798d68. I now have, but currently it doesnt complete all outstanding tasks as I need a bit of help with creating the tests. I'm not sure of the best way to create a test setup that has several product orders with the right associations with producers/coordinators. Would that be best to discuss on skype or could you add some suggestions here?

I have added the receival stuff to record that extra information, including UI changes, so you could check those changes. 

Otherwise the producer_mailer and tests need more work.

RohanM commented:
By a quick look that's looking good. For creating test data, could you give me an exact example of what you'd like to create?

pmackay commented:
@RohanM If you check https://github.com/folklabs/openfoodnetwork/commit/67b17de695ae000dc81501a0440a1b6e42798d68#diff-2ba0593e5da98495e772f742f9ce7587R37 I'm trying to create an order cycle with at least 2 orders that would have the same produce items in, so I can prove they have been aggregated. I'm having issues with the right way to do the following:

* the current code should create an order associated with the order cycle, but it doesnt
* how to set the order status to complete
* how to associate the supplier with the order (it seems easier to set distributor)

Does that help?

RohanM commented:
Hey @pmackay,

It looks like your code to create the order associated with the order cycle is correct. You shouldn't need the `order.set_order_cycle! order_cycle` line. There are two things I'd look into to diagnose this.

First-off, `let` unlike `let!` is lazy-loading, so the objects are not created until they're referenced. Try experimenting with `let!` or just making sure they're both referenced so that they're loaded when you need them.

The second thing you could check are caching issues. ie. `order_cycle.orders(:reload)`.

This is a little hard for me to troubleshoot since I can't see exactly what it is that's not behaving as you would expect. If you could post a gist with a pattern like "When I execute this line I expect to get this back, but instead I get XYZ", that would be really helpful.

RohanM commented:
Setting order state to complete can be done when the order is created:

```ruby
create(:order, state: 'complete')
```

It can also be done later:

```ruby
order.finalize!
```

RohanM commented:
And wrt. associating suppliers with an order, one order does not have a specific supplier, since it can contain products from many different suppliers. Instead, each product in an order has one supplier. A list of all suppliers for an order could be found like this:

```ruby
order.line_items.map { |li| li.product.supplier }.uniq
```

kirstenalarsen commented:
removing tag 'aus review' until it's back this way . .

kirstenalarsen commented:
Hey @pmackay any update on this? Just trying to get a read on when we're likely to need to deal with it next?

pmackay commented:
Keen to progress but this week been too busy. Hopefully next week.

pmackay commented:
@RohanM I've updated the tests a bit, could you take a quick look please?

RohanM commented on app/controllers/admin/order_cycles_controller.rb:
```diff
@@ -70,6 +69,12 @@ def clone
     end
 
+    # Send notifications to all producers who are part of the order cycle
+    def notifications
```
I would expect an action named "notifications" to display a list of notifications. However, what we're doing here is an action, so I would use a verb to name it, probably `notify_producers`.

RohanM commented on app/jobs/order_cycle_notification_job.rb:
```diff
@@ -0,0 +1,8 @@
+
+OrderCycleNotificationJob = Struct.new(:order_cycle) do
+  def perform
+    @suppliers = order_cycle.suppliers
```
There's no need to use an instance variable `@suppliers` here. A local variable would be more suitable since we don't need the var to be accessible by the class, but given the simplicity I would probably just do this on one line without using a variable.

RohanM commented on app/controllers/admin/order_cycles_controller.rb:
```diff
@@ -70,6 +69,12 @@ def clone
+    # Send notifications to all producers who are part of the order cycle
+    def notifications
+      @order_cycle = OrderCycle.find params[:id]
+      Delayed::Job.enqueue OrderCycleNotificationJob.new(@order_cycle)
```
The usual pattern for displaying the result of an action like this would be to redirect back to the order cycles index page with a flash message like "Notification emails sent", rather than rendering a template.

RohanM commented on app/mailers/producer_mailer.rb:
```diff
@@ -0,0 +1,37 @@
+    #   subject += " for #{@distribution_date}" if @distribution_date.present?
+    # end
+
+    # TODO: what if producer handling orders into multiple order cycles?
```
Is this a situation you envisage happening regularly? I can see the advantage in batching up these into a single email per producer, but I can also see it being tricky coordinating emails across several order cycles that close at different times.

RohanM commented on app/views/producer_mailer/order_cycle_report.text.haml:
```diff
@@ -0,0 +1,28 @@
+Here is a summary of the orders for your products:
+\
+- @line_items.each_pair do |variant, data|
+  #{variant.sku} #{raw(variant.product.supplier.name)} #{raw(variant.product.name)} #{raw(variant.options_text)} (QTY: #{data[:quantity]}) @ #{data[:line_item].single_money} = #{variant.display_amount}
```
I haven't tested this myself, but I'm not sure about the money calculations here. In particular, shouldn't `variant.display_amount` be `data[:line_item].display_amount`? This would be a good one to write a spec for to check if it's doing the right thing.

RohanM commented on app/mailers/producer_mailer.rb:
```diff
@@ -0,0 +1,37 @@
+    # puts @orders.size
+
+    # Create a single flat list of all line items
+    @line_items = @orders.map(&:line_items).flatten
```
This will create an n+1 queries problem - for every order, we'll do a separate SQL query to fetch the line items, and then combine the list in ruby. We can fetch the whole lot in one go like this:

```ruby
Spree::LineItem.where(order_id: @orders)
```

RohanM commented on app/mailers/producer_mailer.rb:
```diff
@@ -0,0 +1,37 @@
+    # end
+
+    # TODO: what if producer handling orders into multiple order cycles?
+    @orders = Spree::Order.complete.not_state(:canceled).managed_by(@producer.owner)
```
It looks like this query will fetch more orders than we want:

- If the owner of the producer owns more producers, then we'll get orders involving those producers, too.
- Orders from any order cycle will be returned, not just the current order cycle.

What we want to fetch (as I understand it) are a list of complete orders from this order cycle, and then pull out all the line items that come from the specified producer. A first attempt at this might look like this:

```ruby
orders = Spree::Order.where(order_cycle: order_cycle).complete
@line_items = orders.map(&:line_items).flatten.select { |o| o.variant.product.supplier == producer }
```

However, fetching line items this way is hugely inefficient, because `map` will trigger a separate SQL query for each order we're looking at (loading the line items), and then `select` will trigger *three* queries for each order, to fetch the variant, then the product, then the supplier. All we care about is getting the line items we want, so we can perform this query purely in SQL and avoid going back and forth between Ruby and SQL.

To do this, I like to start by thinking in terms of the resource I want to get back at the end, ie. line items. If we were to phrase that in terms of the line items we want, I'd say: Line items within the order cycle we're looking at, and with the producer of the product being the producer we're looking at. We can construct this query in SQL like this:

```ruby
Spree::LineItem.
    joins(:order => :order_cycle, :variant => :product).
    where('order_cycles.id = ?', order_cycle).
    where('spree_products.supplier_id = ?', producer)
```

I think this should do the trick, but I haven't tried it myself. Have a play and let me know how you go.

RohanM commented:
Hi @pmackay,

I've written lots of code-specific comments above. I have a couple of more general comments as well:

There's a lot of different parts to get this PR done, and it seems like it'd help to be able to break it down into smaller chunks and get those merged one by one. The code to add receival time and instructions to order cycles would be a natural candidate for this since it's self-contained and can be included without breaking anything else. You could do that by cherry-picking the relevant commits across to another branch and creating a PR from that.

There are a few parts that are missing specs, namely the OrderCycleNotificationJob and the controller action.

I've been looking at remote pairing tools, and found one called MadEye that looks promising. There's info here: https://madeye.io/ . Would you like to give it a shot tomorrow to work through some of the things in this PR?

Sorry for the flood of comments, and thanks for the code. It's appreciated!

pmackay commented:
@RohanM here's the output of testings failing when `config.track_inventory_levels = false` :

```
  1) ProducerMailer should send an email when an order cycle is closed
     Failure/Error: let(:p1) { create(:product, price: 12.34, supplier: s1) }
     RuntimeError:
       Cannot set on_hand value when Spree::Config[:track_inventory_levels] is false
     # ./spec/mailers/producer_mailer_spec.rb:8:in `block (2 levels) in <top (required)>'
     # ./spec/mailers/producer_mailer_spec.rb:13:in `block (2 levels) in <top (required)>'
```

kirstenalarsen commented:
@pmackay is this actually aus review or you and @RohanM have discussed already?

pmackay commented:
Removed for now.

RohanM commented:
Just left a comment on the latest commit. In general, looking good!

lin-d-hop commented:
@pmackay 
Just talking with Nick and it would be good if the itemised breakdown in the email also included a total. Would this be ok to add?
The idea is to make invoicing even easier for producers. 
Thanks.

pmackay commented:
@lin-d-hop first I would really like to just get this merged :)

pmackay commented:
@RohanM would appreciate a review and see if this can be merged. Have added:

* permissions checking
* insertion of receival time/instructions

RohanM commented on spec/mailers/producer_mailer_spec.rb:
```diff
@@ -0,0 +1,68 @@
+require 'spec_helper'
+require 'yaml'
```
Just curious, is this actually needed?

RohanM commented:
Hey Paul,

I had a scan over this and it's looking really good. I'll add it to our list of things to build and test. We have quite a backlog, but we'll aim to merge it as soon as we can. To speed up that process could you run all the specs and play around with it yourself to make sure everything's working as you would like?

sstead commented:
Testing note:
![image](https://cloud.githubusercontent.com/assets/10843056/10930971/7a99bf8c-8317-11e5-9be9-a74344992ecb.png)
Also for reference, the email is being sent to the Owner, the Contact, but not the Manager.

sstead commented:
It's not always multiplying by 2...
![image](https://cloud.githubusercontent.com/assets/10843056/10931130/f0c5720e-8318-11e5-85d2-82b286fbaa9e.png)

Also,
![image](https://cloud.githubusercontent.com/assets/10843056/10931120/d9df6e78-8318-11e5-8dcc-aebf57eead76.png)

And, 
![image](https://cloud.githubusercontent.com/assets/10843056/10931184/810d1e84-8319-11e5-93a3-201c3e171051.png)

sstead commented:
Testing: @RohanM 
Yay- the correct quantities and $ values are showing. The variants are also distinguishable now too. Still not holding the instructions in the OC interface (people will complain that this isn't working if disappears after they type it in).

Also I'd like to suggest that if possible we simplify the email....  I'm worried we'll get requests down the track for it to be rephrased because parts won't be true for different enterprises (e.g. some enterprises pick up stock from their producers).

Could we merge the two fields below into one. Just call it 'stock pickup/delivery instructions'.
![image](https://cloud.githubusercontent.com/assets/10843056/11085210/d38bf936-8897-11e5-8249-7c68359a25b6.png)
Then, individual enterprises could put as much information in here as they want.
Then the template email could be made more generic- suggested simplification of email below.

Dear <Producer Name>,

We now have all the consumer orders for next food drop. 
Stock pickup/delivery instructions:
<  Stock pickup/delivery instructions   >

Orders summary
==============
Here is a summary of the orders for your products:

 - Item 1
- Item 2
- ...

Thanks and best wishes, 
<Hub Name>
<Hub Address>
<Hub Phone number>
<Hub Email>

NickWeir63 commented:
thanks @sstead - yes this sounds good to me

RohanM commented:
Those changes are in a branch now, producer-emails-simplify. Should get it tested and pushed soon!

NickWeir63 commented:
I just put a test order through staging 2 and then 'notified producers' and the email came through perfectly.  Thanks very much

sstead commented:
@RohanM Testing the "Merge branch 'fix-delayed-job' into combined/producer-emails-simplify_fix-delayed-job"
- The notify emails are being sent correctly- with new layout.
- Order confirmation emails are also going fine.
- The text is still not holding in the text boxes in the order cycle set up. But I suppose that can be fixed later.

- In BOM. Some of my products had an error in the Weight/volume column.
- This is how the products were set up:
![image](https://cloud.githubusercontent.com/assets/10843056/11522726/973eea88-990e-11e5-8fe0-126d87641cba.png)
- This was my view of the error (Note the Weight/Volume field is collapsed). I can't see what the error is. 
![image](https://cloud.githubusercontent.com/assets/10843056/11522744/c1b30902-990e-11e5-9577-f45ca61d0a81.png)
- Only once I expanded the weight/volume field could I see where the error was:
![image](https://cloud.githubusercontent.com/assets/10843056/11522747/ca0117fc-990e-11e5-8687-28dc50fe4034.png)
- When I tried to click in the red box and type 1, i was not able to type anything. I think the error is relating to how the product is set up- ideally it shouldn't have been allowed to be set up as it is.
